### DISCUSSION TITLE
How to use pytest with a legacy in-house test framework

### DISCUSSION THREAD

**Discussion by danieljfarrell** [pytest-dev/pytest#10649] [Q&A]
I use pytest for all my new projects, but I'm working with a large python 3 project that implemented its own test framework; it does **not** inherit from unittest, it's pure custom python. It follows familiar idioms like setup/teardown, test are class based with assert methods e.g. `self.assert_true` etc.

I'd like to use pytest as a test runner for the legacy tests and start to write new tests using pytest.

How can I run my old tests with pytest? Is there any documentation about this use case? Or has someone done something similar before?

**Comment by arthuRHD** [comment]
Hi @danieljfarrell,

It depends whether or not your "custom" test framework supports external runners. But, you might want to migrate your entire test suites in pytest. Here's why:

1. Less maintainance due to complying to pytest guidelines
2. What if you find a working solution with the actual version of pytest but the next one has breaking changes ?
3. Based on what you describe, it might not be that difficult for pytest to work with your test architecture.

**Comment by RonnyPfannschmidt** [comment]
that strongl depends on the test framework,

if your setup/tearown is in line enough with what pytest expects, it may be possible to just run things (with the caveat of maybe needing to enforce collection of things that have a __init__)


else you pretty much have to implement own collection and own test item classes (just like the builtin unittest plugin does)

**Reply by nicoddemus** [reply]
Indeed the builtin [unittest](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/unittest.py) plugin would be great place to start.

Also searching for `framework` in https://docs.pytest.org/en/stable/reference/plugin_list.html might yield plugins which integrate with external frameworks.
